core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -229,6 +278,74 @@ protected final void requestMessagesFromDeframer(int numMessages) {
     }
   }
 
+  /**
+   * Event handler to be called by the subclass when a number of bytes are being queued for sending
+   * to the remote endpoint. May call back the listener's {@link StreamListener#onReady()}
+   * handler if appropriate.
+   *
+   * @param numBytes the number of bytes being sent.
+   */
+  protected final void onSendingBytes(int numBytes) {
+    boolean shouldNotify = false;
+    synchronized (onReadyLock) {
+      shouldNotifyOnReady = true;
+      numSentBytesQueued += numBytes;
+      shouldNotify = shouldNotifyOnReady();
+    }
+    if (shouldNotify) {
+      listener().onReady();
```

Comment:
Hmm... We can't actually call listener().onReady() here, as we are only supposed to call the listener from the transport thread. Nonetheless, it seems it may be safe from a threading perspective, but we would need to update documentation. There's also the possibility of recursive loops (application sends -> onReady() -> application sends -> onReady()) and application code may not be reentrant.

Would it be better to call onSendingBytes() within the event loop? However, it is nice to be able to know during writing that the application could continue sending, which would reduce latency.

Or maybe we should return a boolean for whether the buffer is full? We would only call onReady() if we returned that the buffer filled. If we had a method isReady(), that would be the Netty behavior, but I'm not sure it applies as well for us.

:-/ all around.